### PR TITLE
Feature/advanced autosuggest

### PR DESCRIPTION
**Pull request by Adamik10**
#### Link to issue
https://reload.atlassian.net/browse/DDFSOEG-144
image

#### Description
This PR will adds the rest of the functionality to our autosuggest that was not included in the basic version:
- updates the DOM structure and classes based on the latest version of the design system npm package
- uses useText() from [#27](https://github.com/danskernesdigitalebibliotek/dpl-react/pull/27) instead of prop drilling. 
- adds material covers 
- if there are no materials, don't show the material suggestion part of the dropdown at all
- fixes the problem with wrong order of items when using down key even if work suggestions came before position 7, 8, 9
- doesn't call the API service on change of the highlighted item, but still updates the input field  
- fixes typescript TODO errors
- simulates form submission when the user selects an item using enter

#### Screenshot of the result
image

#### Checklist
- [x] My complies with [our coding guidelines](../docs/code_guidelines.md).
- [x] My code passes our static analysis suite. If not then I have added a comment explaining why this change should be exempt from the code standards and process.
- [x] My code passes our continuous integration process. If not then I have added a comment explaining why this change should be exempt from the code standards and process.

#### Additional comments or questions
What do you think of the implementation?

### CODE REVIEW

**Review by spaceo — CHANGES_REQUESTED** [review]
Great Adam ;) Have some requests..

File: src/components/autosuggest-material/autosuggest-material.tsx [hunk 1]
```diff
@@ -57,13 +56,13 @@ const AutosuggestMaterial: React.FC<AutosuggestMaterialProps> = ({
...
-                    <div className="text-body-medium-medium autosuggest__item--materials__item__content--info__title">
+                  <div className="autosuggest__info">
+                    <div className="text-body-medium-medium autosuggest__title">
                       {item.fullTitle}
                     </div>
-                    <div className="text-body-small-regular autosuggest__item--materials__item__content--info__author">
+                    <div className="text-body-small-regular autosuggest__author">
                       {item.creators.map((author) => {
-                        if (numberOfAuthors < 3) {
+                        if (numberOfAuthors <= 3) {
```
[the remark this answers is not in the document]
Comment (Adamik10) [reply src/components/autosuggest-material/autosuggest-material.tsx]: dealio :)

File: src/components/autosuggest-material/autosuggest-material.tsx [hunk 2]
```diff
@@ -19,6 +19,10 @@ const AutosuggestMaterial: React.FC<AutosuggestMaterialProps> = ({
   highlightedIndex,
   textDataLength
 }) => {
+  if (materialData.length < 1) {
+    return <div />;
```
[the remark this answers is not in the document]
Comment (Adamik10) [reply src/components/autosuggest-material/autosuggest-material.tsx:27]: hahahaha gotcha ;)

File: src/components/cover/cover.tsx [hunk 3]
```diff
@@ -19,10 +19,16 @@ export const Cover = ({
   tint,
   materialId
 }: CoverProps) => {
+  let dataSize: CoverProps["size"];
```
[the remark this answers is not in the document]
Comment (Adamik10) [reply src/components/cover/cover.tsx:22]: @spaceo our eslint is not happy about that solution:
image

File: src/apps/search-header/search-header.tsx [hunk 4]
```diff
@@ -85,29 +91,67 @@ const SearchHeader: React.FC<SearchHeaderProps> = ({
...
+
   function handleHighlightedIndexChange(
     changes: UseComboboxStateChange<Suggestion>
   ) {
+    const { selectedItem, highlightedIndex, type } = changes;
     if (
-      changes.selectedItem === undefined ||
-      changes.selectedItem === null ||
-      changes.highlightedIndex === undefined
+      selectedItem === undefined ||
```
[the remark this answers is not in the document]
Comment (Adamik10) [reply src/apps/search-header/search-header.tsx:109]: Yes :)

File: src/apps/search-header/search-header.tsx [hunk 5]
```diff
@@ -85,29 +91,67 @@ const SearchHeader: React.FC<SearchHeaderProps> = ({
...
   }
 
-  // here we get all downshift properties for the dropdown that we will need
+  function handleInputValueChange(changes: UseComboboxStateChange<Suggestion>) {
+    const { inputValue, type } = changes;
+    if (inputValue === undefined) {
+      return;
+    }
+
+    if (type === "__input_change__") {
```
[the remark this answers is not in the document]
Comment (Adamik10) [reply src/apps/search-header/search-header.tsx:144]: Done :)

File: src/apps/search-header/search-header.tsx [hunk 6]
```diff
@@ -85,29 +91,67 @@ const SearchHeader: React.FC<SearchHeaderProps> = ({
...
+      setQ(inputValue);
+      setQWithoutQuery(inputValue);
+      return;
+    }
+    setQWithoutQuery(inputValue);
+    if (type === "__controlled_prop_updated_selected_item__") {
```
[the remark this answers is not in the document]
Comment (Adamik10) [reply src/apps/search-header/search-header.tsx:150]: done :)

File: src/components/autosuggest-material/autosuggest-material.tsx [hunk 7]
```diff
@@ -19,51 +20,46 @@ const AutosuggestMaterial: React.FC<AutosuggestMaterialProps> = ({
...
+      <li className="autosuggest__divider" />
       <li>
-        <hr className="autosuggest__divider" />
-      </li>
-      <li>
-        <ul className="autosuggest__item--materials">
+        <ul className="autosuggest__materials">
           {/* eslint-disable react/jsx-props-no-spreading */}
           {/* The downshift combobox works this way by design */}
-          {materialData.map((item, index) => {
```
[the remark this answers is not in the document]
Comment (Adamik10) [reply src/components/autosuggest-material/autosuggest-material.tsx:31]: on line 50

File: src/components/autosuggest-material/autosuggest-material.tsx [hunk 8]
```diff
@@ -19,51 +20,46 @@ const AutosuggestMaterial: React.FC<AutosuggestMaterialProps> = ({
...
+          {materialData.map((item, incorrectIndex) => {
```
[the remark this answers is not in the document]
Comment (Adamik10) [reply src/components/autosuggest-material/autosuggest-material.tsx:36]: Good point :)

File: src/components/autosuggest/autosuggest.tsx [hunk 9]
```diff
@@ -66,18 +46,12 @@ export const Autosuggest: React.FC<AutosuggestProps> = ({
 
         {isLoading && <div>Loading</div>}
 
-        {data && data.suggest.result.length < 1 && null}
-
-        {data && status === "success" && isOpen && (
+        {originalData !== undefined && status === "success" && isOpen && (
```
[the remark this answers is not in the document]
Comment (Adamik10) [reply src/components/autosuggest/autosuggest.tsx:49]: right :)

File: src/components/autosuggest-material/autosuggest-material.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (spaceo) [reply src/components/autosuggest-material/autosuggest-material.tsx]: Don't see any changes?

File: src/apps/search-header/search-header.tsx [hunk 10]
```diff
@@ -80,29 +96,66 @@ const SearchHeader: React.FC<SearchHeaderProps> = ({
...
+    if (window.top) {
+      window.top.location.href = `${baseUrl}?q=${params}`;
+    }
+  }
+
   function handleHighlightedIndexChange(
     changes: UseComboboxStateChange<Suggestion>
   ) {
+    const { selectedItem, highlightedIndex, type } = changes;
+    if (!selectedItem || !highlightedIndex) {
```
Comment (spaceo) [inline src/apps/search-header/search-header.tsx:113]: Maybe a small comment? (if it makes sense)

File: src/apps/search-header/search-header.tsx [hunk 11]
```diff
@@ -80,29 +96,66 @@ const SearchHeader: React.FC<SearchHeaderProps> = ({
...
+      return;
+    }
     if (
-      changes.selectedItem === undefined ||
-      changes.selectedItem === null ||
-      changes.highlightedIndex === undefined
+      type === downshiftEventTypes.item_mouse_move ||
```
Comment (spaceo) [inline src/apps/search-header/search-header.tsx:117]: Here you can also use this construct:
```ts
[
  downshiftEventTypes.item_mouse_move,
  downshiftEventTypes.menu_mouse_leave
].includes(type)

File: src/apps/search-header/search-header.tsx [hunk 12]
```diff
@@ -80,29 +96,66 @@ const SearchHeader: React.FC<SearchHeaderProps> = ({
...
     }
+    const arrayIndex: number = highlightedIndex;
+    const currentlyHighlightedObject = orderedData[arrayIndex];
+    const currentItemValue = determinSuggestionType(currentlyHighlightedObject);
+    if (type === downshiftEventTypes.controlled_prop_updated_selected_item) {
+      manualRedirect(currentItemValue);
+    }
+    if (
+      type === downshiftEventTypes.input_keydown_arrow_down ||
+      type === downshiftEventTypes.input_keydown_arrow_up
```
Comment (spaceo) [inline src/apps/search-header/search-header.tsx:134]: Here you can also use this construct:
```ts
[
  downshiftEventTypes.input_keydown_arrow_down,
  downshiftEventTypes.input_keydown_arrow_up
].includes(type)

**Review by spaceo — APPROVED** [review]

File: src/components/cover/cover.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (Adamik10) [reply src/components/cover/cover.tsx]: Aaaah I see!! ❤️ I was thinking inside my box. Corrected!

File: src/apps/search-header/search-header.tsx (hunk 11 above)
Comment (Adamik10) [reply src/apps/search-header/search-header.tsx:117]: Do you think that's cleaner?

File: src/apps/search-header/search-header.tsx (hunk 12 above)
Comment (Adamik10) [reply src/apps/search-header/search-header.tsx:134]: Do you think that's cleaner? What about consistency then?

File: src/apps/search-header/search-header.entry.tsx [hunk 13]
```diff
@@ -1,22 +1,23 @@
...
+import SearchHeader from "./search-header";
+
+export interface SearchHeaderProps {
+  searchHeaderUrlText?: string;
+  altText?: string;
+  inputPlaceholderText?: string;
+  stringSuggestionAuthorText?: string;
+  stringSuggestionWorkText?: string;
+  stringSuggestionTopicText?: string;
+}
```
Comment (spaceo) [inline src/apps/search-header/search-header.entry.tsx:12]:
```suggestion
etAlText: string;
}
```

**Review by sinejespersen — APPROVED** [review]

File: src/apps/search-header/search-header.entry.tsx (hunk 13 above)
Comment (Adamik10) [reply src/apps/search-header/search-header.entry.tsx:12]: Yesss!!

File: src/apps/search-header/search-header.entry.tsx [hunk 14]
```diff
@@ -1,22 +1,23 @@
...
-    stringSuggestionTopicText={stringSuggestionTopicText}
-  />
-);
+  searchHeaderUrlText = "https://bibliotek.dk/search",
+  altText = "search icon",
+  inputPlaceholderText = "Search here",
+  stringSuggestionAuthorText = "author",
+  stringSuggestionWorkText = "work",
+  stringSuggestionTopicText = "topic"
+}) => <SearchHeader />;
```
[the remark this answers is not in the document]
Comment (Adamik10) [reply src/apps/search-header/search-header.entry.tsx:21]: 👍

**Pull request merged**